keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1197 @@
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- Enable cluster autoscaling when pods use resource claims, with correct
+  decisions and changing the cluster size by more than one node at a time.
+
+- Support node-local resources. Adding or removing nodes has no effect
+  on network-attached resources and therefore CA does not need to (and cannot)
+  simulate them.
+
+- Allow DRA driver developers to provide a user experience that is similar to
+  the one possible without semantic parameters. Ideally, users should not notice
+  at all that a driver is using semantic parameters under the hood.
+
+### Non-Goals
+
+- Support semantic class parameters. They remain opaque and can be used to
+  define configuration of resources, but must not affect where a claim can be
+  allocated.
+
+- Scheduling performance is expected to become better compared to using the
+  PodSchedulingContext. However, this is not the reason for this KEP.
+
+
+## Proposal
+
+### Publishing node resources
+
+The kubelet publishes NodeResourceSlices with content provided by the different
+drivers running on the node. Access control through the node authorizer ensures
+that one kubelet is not allowed to create or modify NodeResourceSlices
+belonging to other nodes.
+
+NodeResourceSlices are published separately for each driver, using a version of
+the API that matches what that driver is using. This implies that kubelet, DRA
+driver, and apiserver must all support that same API version. It might be
+possible to support version skew (= keeping kubelet at an older version than
+the control plane and the DRA drivers) in the future, but currently this is out
+of scope.
+
+Embedded inside the NodeResourceSlices is the list of resources managed by the
+driver. Typically there is a single object per driver, but this could also get
+spread out across multiple if a single object would be too large.
+
+Here is an example where the driver provides one discrete card which has
+a certain amount of compute units and RAM that can be reserved:
+
+```
+kind: NodeResourceSlice
+apiVersion: resource.k8s.io/v1alpha2
+...
+nodeName: worker-1
+driverName: cards.dra.example.com
+resources:
+- name: CardA
+  attributes:
+  - name: meta
+    metadata:
+      versions:
+      - name: API
+        version: "v3.2.1" # Must comply with semver v2.0.0.
+      labels:
+      - name: SupportsPowerSaving
+      - name: Color
+        value: "Red" # arbitrary string
+      values:
+      - name: LengthInCentimeters
+        quantity: 100 # int64
+  - name: RAM
+    counter:
+      count: 1000000 # int64, could also be a resource.Quantity for readability
```

Comment:
@johnbelamaric: in your proposal you had a `block` value. I left that out because there are two other ways of achieving "allocation in blocks":

- The "count" (here) and "required" (in the parameters) values can divided by the block size. So if there are 128 items which can be allocated in chunks of 32, "count" becomes 4.

- The count remains at it is, and the translation from vendor CRD to ResourceClaimParameters rounds up the "required" value. This works less well when users are allowed to create their own ResourceClaimParameters because they might get it wrong. The advantage is that if the block size somehow depends on some other parameters, then the driver can round accordingly.

Regarding `resource.Quantity`: I prefer int64 because a quantity could also be fractional, which doesn't make sense here. I just wanted to mention the option.

Reply:
I think we want to allow users to specify things in natural units, rather than having to know the block size. They should say, "my workload needs 2.5Gi of memory" and don't need to know that because of the particular card/driver that was selected, this means they will be allocated 3Gi. If we just change the "count" then the coupling is tighter and the user has to know more.

That's the reason for the block. But it doesn't need to be a separate thing than "counter". The counter model just needs to be able to specify a block size for consumption, and default it to "1".

I didn't do it that way though, for the reason you state - I wanted 'counter' to be int64 and block to be resource.Quantity.

Reply:
> I think we want to allow users to specify things in natural units, rather than having to know the block size. 

That's what they can do in the vendor CRD. I expect that in almost all cases that will be a better user-facing API that directly letting users create a ResourceClaimParameter.